src/cmd/hitwriter.cpp

```diff
@@ -1,31 +1,118 @@
 #include "hitwriter.h"
 
+#include <algorithm>
+#include <iterator>
 #include <ostream>
 
+
 void nullWriter(void* userData, const LG_SearchHit* const) {
   HitCounterInfo* hi = static_cast<HitCounterInfo*>(userData);
   ++hi->NumHits;
 }
 
 void writeHit(HitWriterInfo* hi, const LG_SearchHit* const hit) {
+  ++hi->NumHits;
   const LG_PatternInfo* info = lg_pattern_info(hi->Map, hit->KeywordIndex);
 
   hi->Out << hit->Start << '\t'
           << hit->End << '\t'
           << reinterpret_cast<uint64_t>(info->UserData) << '\t'
           << info->Pattern << '\t'
-          << info->EncodingChain << '\n';
-
-  ++hi->NumHits;
+          << info->EncodingChain;
 }
 
 void hitWriter(void* userData, const LG_SearchHit* const hit) {
   HitWriterInfo* hi = static_cast<HitWriterInfo*>(userData);
   writeHit(hi, hit);
+  hi->Out << '\n';
 }
 
 void pathWriter(void* userData, const LG_SearchHit* const hit) {
   PathWriterInfo* hi = static_cast<PathWriterInfo*>(userData);
   hi->Out << hi->Path << '\t';
   writeHit(hi, hit);
+  hi->Out << '\n';
+}
+
+void writeGroupSeparator(LineContextHitWriterInfo* hi) {
+  // print the group separator
+  if (hi->BeforeContext > 0 || hi->AfterContext > 0) {
+    if (hi->FirstHit) {
+      hi->FirstHit = false;
+    }
+    else {
+      hi->Out << hi->GroupSeparator << '\n';
+    }
+  }
+}
+
+void writeLineContext(LineContextHitWriterInfo* hi, const LG_SearchHit* const hit) {
```

Comment:
I'd add two helper functions here, one to determine the left context and another to determine the right context. The iterators can then be initialized to the correct values directly and those helper functions can be unit-tested directly. That will then make this function shorter.

Reply:
Done. Was already doing that when you made this comment.